swift/common/middleware/versioned_writes/oio_object_versioning.py

```diff
@@ -0,0 +1,738 @@
+            #     self._build_versions_object_name(obj, version))
+            # req.headers['X-Backend-Allow-Reserved-Names'] = 'true'
+
+            resp = req.get_response(self.app)
+            if resp.is_success:
+                resp.headers['X-Object-Version-Id'] = version
+
+            # Well, except for some delete marker business...
+            is_del_marker = DELETE_MARKER_CONTENT_TYPE == resp.headers.get(
+                'X-Backend-Content-Type', resp.headers['Content-Type'])
+
+            if req.method == 'HEAD':
+                drain_and_close(resp)
+
+            if is_del_marker:
+                hdrs = {'X-Object-Version-Id': version,
+                        'Content-Type': DELETE_MARKER_CONTENT_TYPE}
+                raise HTTPNotFound(request=req, headers=hdrs)
+            return resp
+
+    def handle_request(self, req, versions_cont, api_version, account,
+                       container, obj, is_enabled):
+        if req.method == 'PUT':
+            return self.handle_put(
+                req, versions_cont, api_version, account, obj,
+                is_enabled)
+        elif req.method == 'POST':
+            return self.handle_post(req, versions_cont, account)
+        elif req.method == 'DELETE':
+            return self.handle_delete(
+                req, versions_cont, api_version, account,
+                container, obj, is_enabled)
+
+        # GET/HEAD/OPTIONS
+        resp = req.get_response(self.app)
+
+        resp.headers['X-Object-Version-Id'] = resp.headers[VERSION_ID_HEADER]
+        # Check for a "real" version
+        loc = wsgi_unquote(resp.headers.get('Content-Location', ''))
+        if loc:
+            _, acct, cont, version_obj = split_path(loc, 4, 4, True)
+            if acct == account and cont == versions_cont:
+                _, version = self._split_version_from_name(version_obj)
+                if version is not None:
+                    resp.headers['X-Object-Version-Id'] = version.internal
+                    content_loc = wsgi_quote('/%s/%s/%s/%s' % (
+                        api_version, account, container, obj,
+                    )) + '?version-id=%s' % (version.internal,)
+                    resp.headers['Content-Location'] = content_loc
+        symlink_target = wsgi_unquote(resp.headers.get('X-Symlink-Target', ''))
+        if symlink_target:
+            cont, version_obj = split_path('/%s' % symlink_target, 2, 2, True)
+            if cont == versions_cont:
+                _, version = self._split_version_from_name(version_obj)
+                if version is not None:
+                    resp.headers['X-Object-Version-Id'] = version.internal
+                    symlink_target = wsgi_quote('%s/%s' % (container, obj)) + \
+                        '?version-id=%s' % (version.internal,)
+                    resp.headers['X-Symlink-Target'] = symlink_target
+        return resp
+
+
+class ContainerContext(ObjectVersioningContext):
```

Comment:
We should overwrite the class instead of copying it. The `handle_delete` and` enable_versioning` methods are not modified.